### DISCUSSION TITLE
Get jar with all transitive dependencies

### DISCUSSION THREAD

**Discussion by nvysyaraju-sift** [nielsbasjes/yauaa#883] [Q&A] [answered]
Hello,

I am trying to test the library's functionality by writing an inline java UDF, I used `yauaa-7.10.0.jar` but it shows that some class are not found when using the UDF to parse UA strings. So I was trying to build a fat jar but facing a lot of maven-toolchain issues, so thought of asking if there is a place I could find it. I used the `yauaa-snowflake-7.10.0-udf.jar` for my inline UDF but while compiling the create function, it is giving me `Failed to compile inline Java code due to a bug in the JDK.` in snowflake. 

Basically this is the code I am trying to test, hoping that it will give me some basic attributes like Browser, OS, Device type etc
```
import java.util.*;
import nl.basjes.parse.useragent.UserAgentAnalyzer;
import nl.basjes.parse.useragent.UserAgent;

public class yauaa{

    public static Map<String, String> parseUserAgent(String userAgent) {
    
    UserAgentAnalyzer uaa = UserAgentAnalyzer
            .newBuilder()
            .hideMatcherLoadStats()
            .withCache(10000)
            .build();
    UserAgent agent = uaa.parse(userAgent);
    Map<String,String> attributes = new HashMap<String, String>();

    attributes.put("Browser", agent.getValue("AgentName"));
    attributes.put("Browser_Version", agent.getValue("AgentVersion"));
    attributes.put("Device_Type", agent.getValue("DeviceClass"));
    attributes.put("OS", agent.getValue("OperatingSystemName"));
    attributes.put("OS_Version", agent.getValue("OperatingSystemVersion"));
    
    return attributes; 
    }   
}
```
Can you help me get a fat jar to test the code or please verify if the above agent object will give results for attributes like WebAppName along with common attributes (Browser, OS, Device type etc) ?

**Comment by nielsbasjes** [comment, marked as answer]
To me it is very unclear what you are doing here.
You use the snowflake udf jar and then show java code and ask for the transitive dependencies.
You also mention toolchain problems which are purely your build software.

My current impression is that your question is not a Yauaa problem but a problem with your project.

I have made a very simple demo project part of the integration tests that you can use as a reference to compare
https://github.com/nielsbasjes/yauaa/tree/main/analyzer/src/it/JavaPlain/Java17

For you I think just copying this example and remove the toolchain plugin from the pom.xml should give you a basic starting point.

**Reply by nvysyaraju-sift** [reply]
Apologies for the confusion here, I was following up from my email to you so I haven't provided enough context.

To be clear, I am trying the yauaa parser by creating a snowflake in-line UDF, the java code I shared above is used to create UDF with `yauaa-7.10.0.jar` is used through import statement in UDF. 

While using the created UDF, I am facing some java.lang.NoClassDefFoundError exceptions (eg: org/apache/logging/log4j/LogManager), so I figured I would try to build a fat jar from the yauaa repo after cloning it to my local computer but I was facing some `maven-toolchain-plugin` errors where it could not find matching ToolChain for different versiosn of JDK such as 8, 11 etc.

So I was looking for a fat jar in case if it got already published to maven central.

But to circle back, my main goal here is to test the Yauaa parser's functionality by writing the same code I shared in the initial message above, thank you for pointing the demo project, in demo.java file, it initializes the analyzer object and then the parse() method in that should return a `UserAgent` agent object, from which we can extract individual attributes separately right?

**Reply by nielsbasjes** [reply]
To have a working build environment easily you can use the provided Docker image. See https://yauaa.basjes.nl/developer/building/
This includes all the tools and configs needed to build and test yauaa from source (including the toolchain setup).

If you look at the pom.xml for the snowflake udf you'll find the setup to create a jar with all dependencies included.

**Reply by nvysyaraju-sift** [reply]
On a different note, as I mentioned previously, I used pre-compiled java UDF to referred here (https://yauaa.basjes.nl/udf/snowflake/), it used this [class and method](https://github.com/nielsbasjes/yauaa/blob/main/udfs/snowflake/src/main/java/nl/basjes/parse/useragent/snowflake/ParseUserAgent.java). I parsed UA strings using the handler of this pre-compiled UDF and got good results, now in my project after adding yauaa as dependency and importing it, should I replicate same code that is given in this file (https://github.com/nielsbasjes/yauaa/blob/main/udfs/snowflake/src/main/java/nl/basjes/parse/useragent/snowflake/ParseUserAgent.java) to generate good results for UA like I observed with the UDF? Apologies if these are redundant questions, just not much familiar with these stuff.
